### PR TITLE
update Dockerfile to run as non-root

### PR DESCRIPTION
**Pull request by guusvw**
solves #82 
Signed-off-by: Guus van Weelden <guus.vanweelden@moia.io>

### CODE REVIEW

File: .dockerignore [hunk 1]
```diff
@@ -1,2 +1,3 @@
 .git
+.travis.yml
```
Comment (hutchic) [inline .dockerignore]: I think we still want this

Comment (guusvw) [inline .dockerignore]:
```suggestion
```

Comment (guusvw) [reply .dockerignore]: removed it. it was part of the `.dockerignore`, but fine for me ;)

File: Dockerfile [hunk 2]
```diff
@@ -1,13 +1,31 @@
-FROM golang:1.13.0 AS build
+FROM golang:1.13 AS build
```
Comment (hbagdi) [inline Dockerfile]: We would like to have control over the patch version of Go as well.
Not pinning the patch version can result in using an older version of Go due to Docker's cache.

Please revert this change.

File: Dockerfile [hunk 3]
```diff
@@ -1,13 +1,31 @@
...
 WORKDIR /deck
-COPY go.mod ./
-COPY go.sum ./
+COPY . .
```
Comment (hbagdi) [inline Dockerfile]: Please revert this change.
It is intentional that we only first copy over `go.mod` and `go.sum`. 

This gives Docker image caching, meaning that every time we execute a `docker build`, it won't result in a new hash and re-download all the modules in the next step.

File: Dockerfile [hunk 4]
```diff
@@ -1,13 +1,31 @@
...
 
 FROM alpine:3.10
-RUN apk --no-cache add ca-certificates
-WORKDIR /root/
+RUN apk update \
+    && apk upgrade \
+    && apk --no-cache add ca-certificates
+
+ENV USER=appuser
+ENV GROUP=appusers
```
Comment (hbagdi) [inline Dockerfile]: Let's name the user and group as `deck`.

File: Dockerfile [hunk 5]
```diff
@@ -1,13 +1,31 @@
...
+    --no-create-home \
+    --uid "$UID" \
+    "$USER"
+
 COPY --from=build /deck/deck /usr/local/bin
+
+USER $USER
+
 ENTRYPOINT ["deck"]
+CMD ["help"]
```
Comment (hbagdi) [inline Dockerfile]: Why is this changed?

File: Dockerfile (hunk 2 above)
Comment (guusvw) [inline Dockerfile]:
```suggestion
FROM golang:1.13.0 AS build
```

File: Dockerfile [hunk 6]
```diff
@@ -1,13 +1,31 @@
 FROM golang:1.13.0 AS build
 WORKDIR /deck
-COPY go.mod ./
-COPY go.sum ./
+COPY . .
```
Comment (guusvw) [inline Dockerfile]:
```suggestion
COPY go.mod ./
COPY go.sum ./
```

File: Dockerfile [hunk 7]
```diff
@@ -3,11 +3,30 @@ WORKDIR /deck
 COPY go.mod ./
 COPY go.sum ./
 RUN go mod download
```
Comment (guusvw) [inline Dockerfile]:
```suggestion
RUN go mod download
ADD . .
```

File: Dockerfile [hunk 8]
```diff
@@ -7,7 +7,27 @@ ADD . .
...
 
 FROM alpine:3.10
-RUN apk --no-cache add ca-certificates
-WORKDIR /root/
+RUN apk update \
+    && apk upgrade \
+    && apk --no-cache add ca-certificates
+
+ENV USER=appuser
+ENV GROUP=appusers
```
Comment (guusvw) [inline Dockerfile]:
```suggestion
ENV GROUP=deck
```

File: Dockerfile [hunk 9]
```diff
@@ -7,7 +7,27 @@ ADD . .
 RUN CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build -o deck
 
 FROM alpine:3.10
-RUN apk --no-cache add ca-certificates
-WORKDIR /root/
+RUN apk update \
+    && apk upgrade \
+    && apk --no-cache add ca-certificates
+
+ENV USER=appuser
```
Comment (guusvw) [inline Dockerfile]:
```suggestion
ENV USER=deck
```

File: Dockerfile (hunk 5 above)
Comment (guusvw) [reply Dockerfile]: @hbagdi I (personally) like to have a default command as long as I'm not passing anything to the container and the `help` command is a nice default target,  but I could also remove it

File: Dockerfile [hunk 10]
```diff
@@ -7,7 +7,27 @@ ADD . .
...
+    --no-create-home \
+    --uid "$UID" \
+    "$USER"
+
 COPY --from=build /deck/deck /usr/local/bin
+
+USER $USER
+
 ENTRYPOINT ["deck"]
+CMD ["help"]
```
Comment (guusvw) [inline Dockerfile]:
```suggestion
```

**Comment by digikin** [comment]
There is no reason to add all that group and env into the file. You are only copying a binary into the folder.  Have you looked or ran my dockerfile?

File: Dockerfile [hunk 11]
```diff
@@ -7,7 +7,26 @@ ADD . .
...
-RUN apk --no-cache add ca-certificates
-WORKDIR /root/
+RUN apk update \
+    && apk upgrade \
+    && apk --no-cache add ca-certificates
+
+ARG USER=deck
+ARG GROUP=deck
+ARG UID=9999
+ARG GID=9999
```
Comment (guusvw) [inline Dockerfile]: This allows you to set those variables from outside during the build time of the container @digikin

File: Dockerfile [hunk 12]
```diff
@@ -7,7 +7,26 @@ ADD . .
 RUN CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build -o deck
 
 FROM alpine:3.10
-RUN apk --no-cache add ca-certificates
-WORKDIR /root/
+RUN apk update \
+    && apk upgrade \
```
Comment (hbagdi) [inline Dockerfile]: Why is an upgrade necessary here?

File: Dockerfile (hunk 11 above)
Comment (hbagdi) [reply Dockerfile]: I don't think we really had to template these in the first place. This is a self-contained and small Dockerfile.

No strong feelings here so not changes necessary.

**Comment by hbagdi** [comment]
@guusvw Friendly ping.

**Comment by hbagdi** [comment]
#84 implements this.
Thank you for your PR!
